### PR TITLE
feat: added vpa to `avm/res/container-service/managed-cluster`

### PR DESCRIPTION
**Pull request by PixelRobots**
Signed-off-by: PixelRobots <22979170+PixelRobots@users.noreply.github.com>

## Description

Add VPA as it has now gone GA

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.container-service.managed-cluster](https://github.com/PixelRobots/bicep-registry-modules/actions/workflows/avm.res.container-service.managed-cluster.yml)      |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/container-service/managed-cluster/main.bicep [hunk 1]
```diff
@@ -369,6 +369,16 @@ param identityProfile object?
 @description('Optional. Enables Kubernetes Event-driven Autoscaling (KEDA).')
 param kedaAddon bool = false
 
+@description('Optional. Whether to enable VPA add-on in cluster. Default value is false.')
+param vpaAddon bool = false
```
Comment (ilhaan) [inline avm/res/container-service/managed-cluster/main.bicep:373]: @PixelRobots Is this bool redundant? Can we just use the config of `vpaAddonAutoscaling` if a user wants to use this feature or not and set the `enabled` param in `verticalPodAutoscaler`? I am trying to avoid a situation where someone sets `vpaAddonAutoscaling` but does not set `vpaAddon` to true and wonders why the change didn't go through.

File: avm/res/container-service/managed-cluster/main.bicep [hunk 2]
```diff
@@ -628,6 +638,10 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2024-03-02-p
       keda: {
         enabled: kedaAddon
       }
+      verticalPodAutoscaler: {
+        addonAutoscaling: vpaAddonAutoscaling
+        enabled: vpaAddon
```
Comment (ilhaan) [inline avm/res/container-service/managed-cluster/main.bicep:643]: Based on my earlier comment we can set this to something like this: 
```suggestion
        enabled: (vpaAddonAutoscaling == "Enabled") ? 'enabled' : 'disabled'
```
That way this feature can be controlled by one param instead of two.

Comment (PixelRobots) [reply avm/res/container-service/managed-cluster/main.bicep:643]: Looking at the API docs addonAutoscaling also enables vpa add-on but also configures vpa for the AKS addona. You might not always want that so I did it separately.

The API docs say "Whether VPA add-on is enabled and configured to scale AKS-managed add-ons."

I can do some further testing to confirm if you like.

Thoughts?

Comment (ilhaan) [reply avm/res/container-service/managed-cluster/main.bicep:643]: @PixelRobots Not sure I follow. Are you saying that we should add a check to see if addonAutoscaling is enabled before enabling vpaAddon?

Comment (PixelRobots) [reply avm/res/container-service/managed-cluster/main.bicep:643]: I believe they are independent of each other. 

Both will enable vpa. One will also configure the AKS managed addons, think CSI drivers, etc to make use of the vpa.

Comment (ilhaan) [reply avm/res/container-service/managed-cluster/main.bicep:643]: So what happens if I set my params this way: 
```bicep
vpaAddon = false
vpaAddonAutoscaling = 'Enabled'
```
Should a check be added to ensure vpaAddon is `true` before configuring `addonAutoscaling`?

File: avm/res/container-service/managed-cluster/main.bicep [hunk 3]
```diff
@@ -628,6 +638,10 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2024-03-02-p
...
+      }
```
Comment (ilhaan) [inline avm/res/container-service/managed-cluster/main.bicep:644]: Perhaps this should be something like this: 
```suggestion
      verticalPodAutoscaler: {
        addonAutoscaling: (vpaAddon != false) ? vpaAddonAutoscaling : 'Disabled'
        enabled: vpaAddon
      }
```

**Review by ilhaan — CHANGES_REQUESTED** [review]

File: avm/res/container-service/managed-cluster/main.bicep (hunk 3 above)
Comment (PixelRobots) [reply avm/res/container-service/managed-cluster/main.bicep]: Seems addonAutoscaling is in preview, so i have removed it.

File: avm/res/container-service/managed-cluster/main.bicep (hunk 2 above)
Comment (PixelRobots) [reply avm/res/container-service/managed-cluster/main.bicep:635]: Seems addonAutoscaling is in preview, so i have removed it.

**Comment by ilhaan** [comment]
@PixelRobots Thanks for the update. Can you please re-run the pipeline to verify the new updates you pushed are good to go?

**Comment by PixelRobots** [comment]
> @PixelRobots Thanks for the update. Can you please re-run the pipeline to verify the new updates you pushed are good to go?

**Comment by PixelRobots** [comment]
> @PixelRobots Thanks for the update. Can you please re-run the pipeline to verify the new updates you pushed are good to go?

re ran and all working.

**Comment by PixelRobots** [comment]
any update on pushing this change?

**Comment by AlexanderSehr** [comment]
Hey @ilhaan, could you take another look?

**Review by ilhaan — APPROVED** [review]

**Comment by ilhaan** [comment]
@PixelRobots @AlexanderSehr approved and ready to merge.

**Comment by AlexanderSehr** [comment]
@ilhaan & @JPEasier, please note, the module update was **not** published because the WAF-reliability tests did not pass. Please address them as soon as you can and make sure this code makes it into the registry.

**Comment by PixelRobots** [comment]
> @ilhaan & @JPEasier, please note, the module update was **not** published because the WAF-reliability tests did not pass. Please address them as soon as you can and make sure this code makes it into the registry.
> 
> cc: @PixelRobots

I have fixed the module with this pr, apart from the default tests, which I believe should not be tested against for this ps rule. https://github.com/Azure/bicep-registry-modules/pull/3228
